new logger

herbertgithubaccount opened the pull request:
im slowly learning how git actually works. hopefully this time everything is in order.

now with support for child loggers (by specifying module names), and the `writet` method has support for associative arrays

Greatness7 commented:
I'm personally not very keen on the metamethods. 
```lua
if (log.level >= Logger.LEVEL.DEBUG) then end
if (#log >= Logger.LEVEL.DEBUG) then end
```
You only save a few characters, and now the code looks unusual. Conceptually it doesn't make a lot of sense that a logger would have a "length".

The `__call` is less controversial, but it saves even fewer characters and now no longer matches the syntax we/lua use everywhere else (`niNode.new()`, `table.new()`, etc).
```lua
local log = Logger.new("MyMod")
local log = Logger("MyMod")
```

For me, both metamethods just make code less obvious/consistent for little to no gain.

The other changes seem like reasonable improvements. Though we might want to question whether constants should be upper case `LEVEL.DEBUG`, since AFAIK no existing api follows that convention.

C3pa commented:
It might be desirable to make the standard `logger:debug`, `logger:error`, etc. functions serialize tables using `inspect` or `json.encode` . @jhaakma what do you think?

herbertgithubaccount commented:
i got rid of the questionable metamethods and fixed some minor bugs with the `addToMCM` method. i left the constants unchanged for the time being, but am open to changing them to whatever else you prefer (i dont feel strongly one way or the other).

> It might be desirable to make the standard `logger:debug`, `logger:error`, etc. functions serialize tables using `inspect` or `json.encode` . @jhaakma what do you think?

im a bit uneasy about this since it could override any `__tostring` metamethods the user specifies. (although, it would be possible to check before sending tables to json, so maybe this isn't really a problem.) for the time being though, i left the `write` method unchanged.

herbertgithubaccount commented:
is there still interest in this? i could also add in some of the more recent features i've added to my own personal version, such as:

- passing arguments to predefined functions in the logger. .e.g., `log("table = %s", json.encode, tbl)`
- automatically fetching the `modName` and `moduleName` based on the file the mod is created in. i.e., `Logger.new()` would create a new logger, pulling the `modName` from the folder it's called in (using the `name` field of the TOML metadata if appropriate),  and the `moduleName` from the filename/directory. e.g., the code 
```lua
require("logger").new():info("mod initialized")
```
would generate different outputs based on the file that called it:
```
-- called in "mods/herbert100/test_mod/main.lua"
"[test_mod: INFO] mod initialized"
-- called in "mods/no author test/main.lua"
"[no author test: INFO] mod initialized"
-- called in "mods/no author test/my module.lua"
"[no author test (my module): INFO] module initialized"
```

let me know if these are things the mwse team would be interested in or if it would be better to close the PR for now

C3pa commented:
I'd like to see these changes merged into our current logger. Let's discuss the general API first. Then, we can add the documentation. I haven't read the code yet or studied the API in great detail. So, I'll briefly touch on the points from your introductory post:

1. Log level change to number

I like this. I think it's best to leave the comparison with `logLevel` explicit: ``` if myLogger.logLevel >= Logger.LEVEL.DEBUG then```.

I consider the casing for the `Logger.LEVEL` constants a question open for discussion.

2. I like the addition of the child/nested loggers concept. (Looks like this feature is desired, for example, Merlord's [Crafting Framework](https://github.com/jhaakma/crafting-framework/blob/master/Data%20Files/MWSE/mods/CraftingFramework/util/Util.lua)). I am not sure what the following paragraph is describing:
> In addition, the following syntax may be used:
> i. child = log .. "my module" ii.  child = Logger{modName="my mod", "my module"}

C3pa commented:
Also, why was `colors.lua` file changed?

C3pa commented:
3.
I see it implemented as a child of `Dropdown` with predefined `options` that correspond to available log levels. It would accept a few additional fields in its `new` function argument table:
```lua
-- Note the argument is a table for consistency with the rest of the api
msweMCMCategory:addLogOptions({
    logger = mylogger,
    modName = "My logger name",
    variable = mwse.mcm.createTableVariable({ id = "logLevel", table = config }),
    
})
```
Its `new` method can then check if at least one of the `logger` or `modName` can be resolved to a valid logger. It would also implement a method that would roughly do the following:
```lua
function LogOptions:updateLogLevel()
    self.logger:setLogLevel(self.variable.value)
end

function LogOptions:update()
    self:updateLogLevel()
    Parent.update(self)
end
```

herbertgithubaccount commented:
i really like this idea

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,658 @@
+
+
+---@param writeToFile string|boolean
+---@param onlyThisLogger boolean? should we only update this logger? Default: false
```
What do you think of renaming this argument to something like `updateChildren`?

herbertgithubaccount replied:
i think this would be better. i'll make this change later today

C3pa reviewed (commented):
I haven't reviewed this in great depth, so I've got more work to do.

How much chance is there to integrate these changes into the existing logger in a backwards-compatible way? From a glance, the following methods are different/missing:
```lua
Logger:setOutputFile
Logger:setLogLevel
Logger:doLog
Logger.getLogger
```
If I am right, the other methods are compatible or overloaded to work with current logger's api.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,658 @@
+---@field modName string the name of the mod this logger is for
+---@field level mwseLogger.LEVEL
+---@field moduleName string? the module this logger belongs to, or `nil` if it's a general purpose logger
+---@field useColors boolean should colors be used when writing log statements? Default: `false`
```
Can you drop this field in favor of doing `if mwse.getConfig("EnableLogColors") then` directly in code to stay the same as it works now?

herbertgithubaccount replied:
i made this change, as well as the other changes you proposed (except the `parentLoggers` thing since i wasn't sure how to move forward on that one)

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,658 @@
+end
+
+-- redefine it to make it behave consistently with `debug.getinfo`
+log_metatable.__call = Logger.debug
```
yeah

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/colors.lua:
```diff
@@ -0,0 +1,83 @@
+--[[Logging colors.
```
it has been done. i just copypasted the existing `colors.lua` into the new logger folder since i'm still not sure about backwards compatibility (or which folder this logger will sit in once everything is finished)

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,658 @@
+function Logger:error(...)
+    if self.level >= LOG_LEVEL.ERROR then self:write("ERROR", ...) end
+end
+
```
this is a good idea. i'll update it soon

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,658 @@
+    -- if there are already loggers with this `modName`, get the most recent one, and then
+    -- update this new loggers values to those of the most recent logger. (if those values weren't specified in `params`)
+    if loggers[modName] and #loggers[modName] > 0 then
+        local parent = loggers[modName][#loggers[modName]]
```
i'm not opposed to this idea, but wanted to clarify first. my current interpretation is that the loggers don't really have a "parent" and "child" relationship, it's more like they're all "siblings" of each other, grouped together by a common `modName`. so i think maybe the `parentLoggers` terminology might be a bit confusing.

there is currently a `Logger.getLoggers(modName)` method, which would behave similarly to the `log.parentLoggers` field. do you think this is sufficient or should i add the `parentLoggers` field instead?

C3pa replied:
I am ok with the sibling terminology. Also, I think `Logger.getLoggers` is enough for now.

herbertgithubaccount commented:
> I haven't reviewed this in great depth, so I've got more work to do.
> 
> How much chance is there to integrate these changes into the existing logger in a backwards-compatible way? From a glance, the following methods are different/missing (also I checked their usage in [Morrowind-nexus-lua-dump](https://github.com/MWSE/morrowind-nexus-lua-dump/tree/master)):
> 
> ```lua
> Logger:setOutputFile -- No users
> Logger:setLogLevel -- Many users
> Logger:doLog -- Only used in Ashfall once
> Logger.getLogger -- Many users
> ```
> 
> If I am right, the other methods are compatible or overloaded to work with current logger's api.

I can add backwards compatibility for all of those methods, since they have equivalents in the new implementation (e.g. `setLevel` instead of `setLogLevel`.) If going this route, maybe it would make sense to do something like
```lua
Logger.setLogLevel = Logger.setLevel
```
and then mark `setLogLevel` as deprecated?

But my main concern with backwards compatibility stuff is that this logger stores its logging level as an `integer`, while the previous implementation stored it as a `string`. if mods were interfacing with the logging level directly, then that might break quite a few things. Maybe there's a nice backwards compatible way to do this, though. (such as changing the `__index` metamethod so that it returns the `string` version of the logging level if someone asks for `log.logLevel`, which is how the logging level is stored in the old implementation.)

C3pa commented:
I surveyed the Morrowind nexus lua dump. Unfortunately, there are some mods that directly write to the logLevel field. Here are the hits: [logLevel usage.txt](https://github.com/MWSE/MWSE/files/14469043/logLevel.usage.txt). Is it even possible to implement compatibility with those mods and still store the logLevel as a number?

We would need these changes integrated into the existing logger file at `core/lib/logging/logger.lua` to merge this in.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,653 @@
+        params = {}
+    elseif type(params) == "string" then
+        params = {modName=params} 
+
+    -- check if the user typed "Logger:new" instead of "Logger.new"
+    elseif params == Logger then
+        params = params2 or {}
+
```
Let's leave it as it works currently: we accept only calling as `Logger.new`.

herbertgithubaccount replied:
sounds good. i'll revert this change

herbertgithubaccount commented:
> I surveyed the Morrowind nexus lua dump. Unfortunately, there are some mods that directly write to the logLevel field. Here are the hits: [logLevel usage.txt](https://github.com/MWSE/MWSE/files/14469043/logLevel.usage.txt). Is it even possible to implement compatibility with those mods and still store the logLevel as a number?

I think it's possible, it would just require a good deal of care to make sure we cover all the edge-cases. We could use the `__newindex` metamethod to handle converting the string to a number:
```lua
__newindex = function(self, k, v)
        if k == "logLevel" then
            self:setLevel(v)
        else
            rawset(self, k, v)
        end
    end,
```
this leads to the following output:
```lua

log:setLevel("DEBUG")
log:info("my log level is %s", log.level) --> "my log level is 4"
log.logLevel = "INFO"
log:info("my log level is %s", log.level) --> "my log level is 3"
```

In this case, we're being helped by the new logger storing its logging level in the `level` field, while the old logger stores it in the `logLevel` field (since `__newindex` is only triggered if the corresponding key is missing from a table, something I learned the hard way multiple times.)

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,714 @@
+
+
+-- support the old way
+---@deprecated 
+Logger.getLogger = Logger.get
```
I am okay with the switch to table argument for `Logger.get`, but `Logger.getLogger` needs to be backwards compatible with 1 string argument.

herbertgithubaccount replied:
i should have probably clarified this better in the commit, but the `Logger.get` method accepts either a `table` with the appropriate parameters, or a `string` (which is interpretted as the `modName`). so, `Logger.get(modName)` is still valid.

but maybe it's better to have `Logger.get` only accept a `table` argument?

herbertgithubaccount commented:
unfortunately, i think it might be for the best to remove the `msg` and `args` fields from `Logger.Record`, due to some weirdness in lua table-packing. here's an example that illustrates the problem:
```lua
print(string.format("testing %s %s %s", 1, nil, nil)) --> "testing 1 nil nil
print(string.format("testing %s %s %s", table.unpack({1, nil, nil}))) --> Error: bad argument #3 to 'format' (no value)
```

this can introduce errors that are quite hard to debug if the user doesn't know what to look for. basically, the problem is that `{...}` deletes any trailing `nil` values, so `unpack({...})` can result in fewer arguments getting sent to `string.format` than may be expected.

if `...` is never packed into a table, this problem does not occur.

i'd be open to finding a different solution to this problem, but thought it would be important to mention it.

C3pa commented:
That looks like a real problem. What do you think if we make the logger's `format` method not responsible for formatting the message and additional arguments? We could handle the formatting in each logging method (warn, error, etc.). Something like:
```lua
for levelStr, level in pairs(LOG_LEVEL) do
    -- e.g., "DEBUG" -> "debug"
    ---@param self Logger
    Logger[string.lower(levelStr)] = function(self, msg, ...)
        if self.level < level then return end

		if type(msg) == "function" then
			msg = msg()
		end
		msg = tostring(msg):format(...)

        local str = self:format(self:makeRecord(msg, level))
        if self.file then
            self.file:write(str, "\n")
            self.file:flush()
        else
            print(str)
        end
    end
end
```
In such case, the `Logger:format` method would lose a bit of flexibility. But the user still has the ability to customize logging of all the other info `Logger.record` and Logger itself store. In addition, our current logger doesn't expose the formatting of additional args into msg string to the user.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,681 @@
+        hasAuthorName = hasAuthorName or false
+    else
+        local oneDirUp = table.concat({"Data Files", "MWSE", "mods", luaParts[1]}, "\\")
+        local lfs = require("lfs")
```
`lfs` library is already in global namespace.

herbertgithubaccount replied:
I could be misremembering, but I think I added this a few months ago when there were some weird crashes/errors happening from the logger being loaded before `core/initialize.lua` finished running. (I think because some things like the `DependencyManager` wanted a logger.)

But I tested this just now and it seems like things mostly work fine without the `lfs = require("lfs")` code. Although I'm still scared to delete it. I might be thinking of something else though.

herbertgithubaccount commented:
> That looks like a real problem. What do you think if we make the logger's `format` method not responsible for formatting the message and additional arguments? We could handle the formatting in each logging method (warn, error, etc.). 

I think I was able to solve this in the latest release of herbert lib in a way that still allows for the `format` to have total control over how messages are formatted.
Basically, the signature of `Logger:format` is changed to 

```lua
-- default formatter. can be overridden by users
---@param record herbert.Logger.Record
---@param msg string|table|any|fun(...):...
---@param ... string|table|any|fun(...):...
---@return string
function Logger:format(record, msg, ...)
```
and this avoids the problem because the `...` parameters are never packed into a table. This does make things a bit clunkier (in terms of how many arguments are passed to `Logger:format`), but it allows the `format` method to retain all of its flexibility. 

The current implementation also allows for the syntax `log:debug(str, func, arg1, arg2, ...)`, which is currently rendered as `string.format(str, func(arg1, arg2, ...))`. (i.e., you can pass a function that takes in multiple arguments and returns multiple arguments, to make things like `log:debug("tbl = %s, json.encode, tbl)` possible.) It may be possible to still allow this sort of functionality if we extend your code sample to something like
```lua
str = string.format(str, func(arg1, arg2, ...))
self:format(str)
```
but I worry it may be too limiting if `Logger:format` is always passed a string. Since I could see some mods wanting the option to specify custom formatting for the data structures they define in their mods. (Or maybe some people would prefer `string.format(func(), arg1, arg2, ...)`.

But this is a pretty small edge-case, so either way it probably won't make too much of a difference for most people, and I'd ultimately be fine with either option. Just wanted to share some pros and cons.

C3pa commented:
I'm ok if you push the changes so the `Logger:format` works as:
```lua
-- default formatter. can be overridden by users
---@param record herbert.Logger.Record
---@param msg string|table|any|fun(...):...
---@param ... string|table|any|fun(...):...
---@return string
function Logger:format(record, msg, ...)
```

herbertgithubaccount commented:
Thanks for taking a close look at this. I'll merge in the changes from herbert lib after I wrap up my exams. Only a little over a week left!

Properly merging the changes from herbert lib will require some care since I refactored the Logger in herbert lib to use my custom class framework, and I also merged in some functionality from other parts of the library. So I think it would be best to postpone the merge until I can give it the attention and care it requires.

The intricacies of table unpacking definitely make it feel like it's one of lua's sharpest corners. But it's good to know we have a workaround for it.

C3pa commented:
I wish you success on your exams! 

No rush with this PR, we can handle this when you find more free time.

herbertgithubaccount commented:
I updated the logger to pull in new functionality from herbert lib. This ended up being a larger refactor than I was hoping for, and it still needs some more robustness testing. I did, however, do a decent amount of testing with various mods and it seems like the basic functionality doesn't have major problems.

Here are the changes that were imported from herbert lib:
1. Loggers now have a `sharedData` field, which housees the shared state of all loggers associated with a particular mod. This includes:
    - The name of the mod.
    - The mod directory.
    - The current logging level.
    - Various general formatting settings.
 The advantage of this is that it makes it much easier to ensure that all shared fields are synchronized between the loggers, and it also cuts down on memory usage.
2. The `args` parameter has been removed from `Logger.Record`. This is not ideal, but it was done to avoid one of Lua's sharp corners: the fact that variadic arguments to functions can have `nil` entries, but "arrays" cannot.
3. The default formatter has been updated. It now tries to prettyprint things. I'm still on the fence about some of these ideas and would like some feedback on them:
    - tables with no `__tostring` metamethod will have `json.encode` called on them.
    - functions will be called, and passed the appropriate number of arguments.
       - Example: If `f` is defined to accept exactly two arguments, then `log:debug(msg, f, a, b, c)` will reduce to `log:debug(msg, f(a, b), c) with `f(a, b)` being computed ONLY if the logging level is appropriate.
      - A similar feature was present in the previous version, but the main difference is that functions can now appear anywhere in the parameter list. E.g., `log:debug(msg, x, y, f, a, b, c)` now becomes `log:debug(msg, x, y, f(a, b), c)`.
      - This might be a footgun though, so feedback would be appreciated.
4. The algorithm for determining the appropriate mod directory of a logger has been updated. The goal is to make it better handle the "logger factories" that appear in several mods.
5. A new formatting option has been added: `abbreviateHeader`. This will shorten the headers that are generated when writing log messages. 
    - Example: 
        - unabbreviated: `[herbert level mod | data/player/derived_data.lua:549 | DEBUG]`
        - abbreviated:     `[herbert level mod | d/p/derived_data:549 | D]`

C3pa commented on misc/package/Data Files/MWSE/core/startLuaMods.lua:
```diff
@@ -13,7 +13,7 @@ local function execLuaMod(runtime)
 		local dependencyManager = dependencyManagerModule.new({
 			metadata = runtime.metadata,
-			logger = logger.new({ name = name, logLevel = "INFO" }),
+			logger = logger.new{ modName = name, level = 3, modDir = runtime.key },
```
It looks a bit weird to pass magic number 3. If info is the default log level can you just drop the `level` parameter?

C3pa commented on misc/package/Data Files/MWSE/core/lib/dependencyManagement/DependencyNotifier.lua:
```diff
@@ -1,29 +1,22 @@
 ---@class DependencyNotifier
 local DependencyNotifier = {
-    ---@type mwseLogger
+    ---@type Logger
```
Let's leave the type name `mwseLogger`.

C3pa commented on misc/package/Data Files/MWSE/core/lib/dependencyManagement/DependencyNotifier.lua:
```diff
@@ -1,29 +1,22 @@
-        self.logger = MWSELogger.new {
-            name = name,
-            logLevel = "INFO"
+function DependencyNotifier.new(cls, e)
```
Why was this change needed?

herbertgithubaccount replied:
That's a good question. I'm not quite sure to be honest. I think I made those changes many months ago when I was trying to figure out why the dependency system was crashing the game with the new logging framework. I'm not entirely sure if these changes were needed but I'll look into it more tomorrow.

The main thing that's changed here is that the newly created logger is being assigned to instances of the class rather than to the class itself (as it was in the previous version). But I don't think I need to rewrite the entire new method to do that. One change I do think is worth keeping is to add in the `moduleName` parameter. I think it makes the dependency logging messages look a bit nicer: 
```
[DependencyManager (Animated Containers Rewritten) | dependencies/main.lua:97  | WARN] Metadata file (Animated Containers Rewritten) found pointing to missing mod files:
```

C3pa replied:
Ok, let's keep the `moduleName` parameter addition.

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,903 @@
+---@param level Logger.LEVEL
+---@param offset integer? for the line number to be accurate, this method assumes it's getting called 2 levels deep (i.e.). the offset adjusts this
+---@return Logger.Record record
+function Logger:makeRecord(level, offset)
```
I think that's a good idea.

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,903 @@
+		-- update the parameters for backwards compatibility
+		params.modName = params.modName or params.name
+		params.level = params.level or params.logLevel
+		params.filePath = params.filePath or params.filepath
```
I wasn't sure if it should be `filepath` or `filePath`. But I think you're right that it's weird to support both. Which do you think is better? I'm slightly leaning towards `filepath` but I'm not sure and could go either way.

C3pa replied:
Yeah `filepath` sounds good.

herbertgithubaccount commented on misc/package/Data Files/MWSE/core/lib/logger/colors.lua:
That's a good idea. Although it now leads to the weird situation where the Logger folder would only contain a file called `init.lua`. Would it be better to move that to `core/lib/logger.lua`? Or perhaps replace `logging/logger.lua`? I would prefer it if the import path was `"logger"` instead of `"logging.logger"`, but it also wouldn't be the end of the world if I had to write an extra `"logging."` when importing the logger.

C3pa replied:
Can you start a discussion on our Discord server about importing the logger as `require("logger")` vs `require("logging.logger")`?

On an unrelated note, I don't see anything wrong with a single file in `core/lib/logger` being called `init.lua`. I use that pattern in my mods.

herbertgithubaccount replied:
I started a thread at https://discord.com/channels/210394599246659585/381219559094616064/1345871494898651146

It seems like the consensus is that `require("logger")` is preferable to `require("logging.logger")`. So I will keep it as is.

herbertgithubaccount replied:
There was more discussion surrounding this after I went to bed. NC suggested moving the `colors` module inside of the `logger` directory as opposed to the `logging` directory. But I have no skin in this game and am fine either way.

I also suggested perhaps making the logger a global variable and Merlord said he wouldn't be opposed to having it be globally available via `mwse.logger`. How would you feel about that idea?

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,1016 @@
+
+
+-- Logging framework.
+---@class Logger : Logger.SharedData
```
Why isn't this class named `mwseLogger`? Isn't the new logger meant to replace the current one?

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,1016 @@
+-- Although for the latter options, it's recommended you use the relevant methods instead.
+---@class Logger.newParams
+---@field logToConsole bool? Should the output be written to the console?
+---@field modName string? the name of the mod this logger is for. will be automatically retrieved if not provided
+---@field modDir string? The directory that this mod operates in. Used to retrieve the `Mod_Info`.
+---@field moduleName string|false? The name of a module that this logger belongs to.
+-- This is useful in cases where one file has several distinct parts. 
+-- The `moduleName` will be displayed next to the name of the mod, in parentheses.
+---@field filePath string? path to the file. will be retrieved if not provided
```
Sorry for not bringing this up before, but this set of arguments sounds like a bit too much: `modName`, `modDir`, `moduleName` and `filePath`.Can't the loggero correctly infer the `modDir` and `filePath`?

herbertgithubaccount replied:
I do agree that it is too much, but my hands are currently a bit forced in terms of the `modDir`, `filepath`, and `modName` being required in the constructed. This is because the dependency management system causes the game to crash without explanation if those parameters aren't explicitly provided. I think it has something to do with the Lua system not being loaded enough to perform the `debug` functionality required to infer those parameters, but I'm not sure exactly what is causing the problem. 

That being said, I could still remove these parameters from the documentation since they won't be needed in 99% of use-cases. 

Alternatively, we could simply make the dependency management system use the `mwse.log` function instead, since I'm not really sure why it needs to use the `mwseLogger`. This would allow us to also remove the code responsible for handling these parameters in the constructor, and it would simplify things a moderate amount. 

Let me know which approach you think is better.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,1016 @@
+---@field includeTimestamp boolean? should timestamps be included in logging messages? Default: `false`
+---@field abbreviateHeader boolean? Should the header messages be abbreviated? Default: `false`.
+---@field formatter Logger.formatter? A way to specify how logging messages should be formatted. A default one will be used if not provided.
+
```
This parameter list is missing `logToConsole`. Please mention the option when writing the documentation.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,1016 @@
+-- The `moduleName` will be displayed next to the name of the mod, in parentheses.
+---@field filePath string? path to the file. will be retrieved if not provided
+---@field level Logger.LEVEL? the log level to set this object to. Default: "LEVEL.INFO"
+---@field writeToFile string|boolean|nil whether to write the log messages to a file, or the name of the file to write to. if `false` or `nil`, messages will be written to `MWSE.log`
```
Can you rename this parameter to `outputFile`, so that it doesn't stray from the already used name?

herbertgithubaccount replied:
I made this change in the `autocomplete` documentation and will soon delete this version once the documentation is more finalized. I also included the `logToConsole` parameter there as well.

C3pa commented on autocomplete/definitions/events/standard/activate.lua:
```diff
@@ -23,14 +23,14 @@ Some examples of when the activate event fires includes:
-			title = "Prevent the player from activating NPCs",
-		},
-	},
+	-- examples = {
+	-- 	["ShowActivationTarget"] = {
+	-- 		title = "Show a message for what the player is activating",
+	-- 	},
+	-- 	["PreventActivatingNPCs"] = {
+	-- 		title = "Prevent the player from activating NPCs",
+	-- 	},
+	-- },
 	links = {
```
You probably didn't want to commit changes to this file.

C3pa commented on tags:
You also probably didn't intend to commit this file.

herbertgithubaccount replied:
yes you're completely right. this is the first time i'm seeing that file and i don't even know how that snuck in or where it came from. sorry about that. i'll try to be more vigilant about these things going forward.

i pushed the most recent commit as a way to save my changes before heading out for a bit, but i should have looked through the changes files a more thoroughly. i fixed the other accidental change as well.

there is still much to be done in terms of documentation though and the current version is still a bit of a rough draft.

C3pa commented on docs/source/guides/logging.md:
````diff
@@ -1,74 +1,240 @@
 
-log:info("This is an info message")
+## Creating a new Logger
+The simplest way to create a new logger is to simply call `Logger.new()` and have all the relevant information be retrieved automatically.
+But it's also possible to pass additional parameters to the `new` function, all of which are optional:
+```lua
+local log = Logger.new{
````
Consider writing this as `mwse.Logger.new`.

C3pa commented on docs/source/guides/logging.md:
````diff
@@ -1,74 +1,240 @@
 ```
+All of the above settings only need to be specified (at most) once, because loggers synchronize their settings.
+This means that, for example, it's enough to specify the `modName` parameter in a single file.
+
+In your main.lua, place the logger creation before other source files are included or required. This is to ensure the logger is created and accessible to these other source files.
+
 
````
Could you add an example of how this new logger is supposed to be used in mods with multiple files? In other words, how it should replace [this ](https://github.com/jhaakma/crafting-framework/blob/master/Data%20Files/MWSE/mods/CraftingFramework/util/Util.lua#L14)construction.

herbertgithubaccount replied:
I should probably make this more clear in the guide somehow, but my primary motivation for making the new logger API was to make it work in multi-file contexts with no extra work required from the user. So the equivalent to the code you linked would be to just write `local log = mwse.Logger.new()` in whatever file you want a logger in. All of the other stuff is taken care of on the back-end side of things, no Logger factories required. 

To give a more concrete example of what this looks like, here are some of the logging messages from More QuickLoot:
```
[More QuickLoot | mod.lua:407 | DEBUG] enabling container
[More QuickLoot | containers/inanimate.lua:154 | DEBUG] updating container state
[More QuickLoot | containers/inanimate.lua:170 | DEBUG] updated lock info! trap = false, lock_lvl = false
[More QuickLoot | gui.lua:999 | DEBUG] doing a quick update!
[More QuickLoot | containers/inanimate.lua:127 | DEBUG] checking if we can enable this container for "tes3reference(\"com_basket_arille's00000000\")"
	relevant handles = { "com_basket_01_ingredien00000001", "com_basket_01_ingredien00000000", "com_basket_arille's00000000" }
	handle: "com_basket_arille's00000000"
[More QuickLoot | containers/inanimate.lua:131 | DEBUG] 	returning....1
[More QuickLoot | gui.lua:915 | DEBUG] got subtitles = {}
```
These logging messages originate from multiple distinct files and no special work was required on my end to display that information in the logging messages (other than making the `Logger` class itself I suppose 😅 )
In particular, all of the loggers in the respective files were just created with a simple `local log = Logger.new()`.

C3pa replied:
That is very nice. The new section in the logging guide describes this clearly.

C3pa commented on misc/package/Data Files/MWSE/core/lib/dependencyManagement/DependencyManager.lua:
```diff
@@ -27,21 +27,18 @@ local util = require("dependencyManagement.util")
 ---@class DependencyManager.new.params
+---@field logLevel Logger.LEVEL?
 ---@field metadata MWSE.Metadata The metadata of the mod using this dependency manager
----@field logger mwseLogger? The logger to use for this dependency manager
----@field logLevel string? The log level to use for this dependency manager if no logger is supplied
+---@field logger Logger? The logger to use for this dependency manager
```
Aren't these annotations out of date?

C3pa commented on misc/package/Data Files/MWSE/core/core tests/common.lua:
```diff
@@ -4,7 +4,7 @@ common.log = require("logging.logger").new({
     logLevel = "TRACE",
     logToConsole = false,
-    includeTimestamp = true,
+    -- includeTimestamp = true,
```
What was the reason this line was commented out? If it doesn't work with `includeTimestamp = true`, feel free to remove it.

herbertgithubaccount replied:
I believe it was commented out because the timestamps were interferring with the tests in `logging.lua`.

C3pa commented on autocomplete/definitions/namedTypes/mwseLogger/new.lua:
```diff
@@ -0,0 +1,25 @@
+			{ name = "modName", 	type = "string", optional = true, description = "The name of MWSE mod associated to this Logger. This will be retrieved automatically if not provided." },
+			{ name = "moduleName", 	type = "string", optional = true, description = "The module this Logger is associated with. This can be useful for distinguishes which parts of your mod produce certain log messages. This will be displayed next to the name of the mod, in parentheses."},
+			
+			{ name = "level", 	     type = "mwseLogger.LOG_LEVEL", optional = true, default = "mwseLogger.LOG_LEVEL.DEBUG", description = "The logging level for all loggers associated to this mod." },
```
Isn't the default log level `mwseLogger.LOG_LEVEL.INFO`?

herbertgithubaccount replied:
Oops. Yes you're right. Thank you for catching that.

C3pa commented on autocomplete/definitions/namedTypes/mwseLogger/new.lua:
```diff
@@ -0,0 +1,25 @@
+		type = "table",
+		optional = true,
+		tableParams = {
+			{ name = "modName", 	type = "string", optional = true, description = "The name of MWSE mod associated to this Logger. This will be retrieved automatically if not provided." },
+			{ name = "moduleName", 	type = "string", optional = true, description = "The module this Logger is associated with. This can be useful for distinguishes which parts of your mod produce certain log messages. This will be displayed next to the name of the mod, in parentheses."},
+			
+			{ name = "level", 	     type = "mwseLogger.LOG_LEVEL", optional = true, default = "mwseLogger.LOG_LEVEL.DEBUG", description = "The logging level for all loggers associated to this mod." },
+			{ name = "logToConsole", type = "boolean", 			    optional = true, default = false, 					     description = "Should the output also be written to the in-game console?" },
+			{ name = "outputFile", 	 type = "boolean|string", 			optional = true, default = false, 					     description = "The path of the output file to write log messages in. This path is taken relative to `Data Files/MWSE/logs/`. If not provided, log messages will be written to `MWSE.log`. If `true`, then the `modDir` will be used as the output path." },
+			
+			{ name = "includeTimestamp", type = "boolean", optional = true, default = true, description = "Should timestamps be included in logging messages? The timestamps are relative to the time that the game was launched." },
+			{ name = "abbreviateHeader", type = "boolean", optional = true, default = false, description = "Should the headers be abbreviated?" },
+			{ name = "formatter", type = "fun(self: Logger, record: mwseLogger.Record, ...: string|any|fun(...): ...): string", optional = true, description = "A custom formatter. This lets you customize how your logging messages are formatted. If not provided, the default formatter will be used." },
```
Can you trim the extra tabs inside these tables?

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/formatters.lua:
```diff
@@ -0,0 +1,266 @@
+	
+	---@param self mwseLogger
+	---@param record mwseLogger.Record
+	formatters.expandAllFunctions = function (self, record, ...)
```
Do you mention this formatter somewhere in the docs?

C3pa replied:
We could have another section on formatters in the logging guide that could make a brief mention of this formatter also.

herbertgithubaccount replied:
Good idea. I added a section on custom formatters to the logging guide.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,866 @@
+
+
+
+local LOG_LEVEL = {
+	NONE  = 0,
+	ERROR = 1,
+	WARN  = 2,
+	INFO  = 3,
+	DEBUG = 4,
+	TRACE = 5
+}
```
Sorry about not mentioning this earlier, but as [G7 remarked](https://github.com/MWSE/MWSE/pull/462#issuecomment-1868127985) at the time this PR was opened, please rename this enumeration to camel case:
```lua
local logLevel = {
        none = 0,
        error = 1,
        -- etc.
}
```

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,866 @@
+---@class mwseLogger : mwseLogger.SharedData
+---@field protected sharedData mwseLogger.SharedData Stores information that's shared between loggers. This should not be accessed directly.
+local Logger = {
+	LOG_LEVEL = LOG_LEVEL,
```
Same as above

herbertgithubaccount replied:
If I rename this to `logLevel` then it would cause a minor backwards compatibility issue due to clashing with the (private) `logLevel` field of the previous logging API. The `logLevel` field is being monitored by the `__index` metamethod of Logger instances, which would mean it gets handled in different ways by Logger objects and the Logger class. (i.e., `Logger.logLevel` would be an enum, but `log.logLevel` would be treated as a field.) This is probably fine, but I wanted to double check first.

C3pa replied:
Yes, we have an inconsistency now. I had a glance at the implementation of our current logger. There, only instances had the `logLevel` field set (and we do have backwards compatibility with that field). The base Logger didn't have this field available. Since we've now decided to expose the enumeration as `mwse.logLevel`, I'd even vote to remove this field from the Logger class. What do you think?

herbertgithubaccount replied:
I think that what you proposed is the best solution. I will remove the enum from the `Logger` class and keep the enum available via `mwse.logLevel`.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,866 @@
+
+-- Make it available globally.
+mwse.Logger = Logger
+mwse.LOG_LEVEL = LOG_LEVEL
```
Same as above

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,866 @@
+-- This function is responsible for updating the `outputFile` field of a logger.
+-- This is only called in one place, but is factored out to help with code readability.
+---@param sharedData mwseLogger.SharedData
+---@param outputFile string|false|nil
```
`---@param outputFile string|boolean|nil`

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,866 @@
+		version = m and m.package and m.package.version
+	end
+	if version then
+		self:write(self:formatter(record, "Initialized version %s.", version))
+	else
+		self:write(self:formatter(record, "Mod initialized."))
+	end
```
Can you use `self:writeRecord` here?

herbertgithubaccount replied:
Good idea.

C3pa commented on misc/package/Data Files/MWSE/core/lib/logger/init.lua:
```diff
@@ -0,0 +1,866 @@
+	end
+end
+
+-- I am a very good programmer.
```
Drop this comment.

C3pa reviewed (changes requested):
Sorry for the bunch of comments. Most are copy-editing and should be simple to resolve.

herbertgithubaccount commented on autocomplete/definitions/namedTypes/mwseLogger.Record/timestamp.lua:
```diff
@@ -0,0 +1,14 @@
+---@field timestamp number|false the timestamp of this message
+return {
+	type = "value",
+	description = "The timestamp that was created when this logging messages was created. This is obtained directly from `socket.gettime()`. \z
+		In particular, it captures the current real-world time, rather than the amount of time since the game launched.that triggered this record to be created.\n\z
+		\n\z
+		Will be `false` if the `Logger.includeTimetamp` is `false`.",
```
Oh wow this one was pretty sloppy. Thank you for catching that.

herbertgithubaccount commented on autocomplete/definitions/namedTypes/mwseLogger/get.lua:
```diff
@@ -0,0 +1,11 @@
+return {
+	type = "function",
+	description = [[Gets a logger associated with a particular `modDir`. You can also filter by `filepath` as well.]],
+	arguments = {
+		{ name = "modDir", type = "string",  description = [[The directory of mod to retrieve the loggers for. This argument corresponds to the `lua-mod` field of the mods metadata file.]] },
+		{ name = "filepath", type = "string",  optional = true, description = [[The filepath to retrieve the logger for. This is relative to `modDir` and corresponds to the `filepath` field of the logger to retrieve.]] }
+	},
+	returns = {
+		{ name = "logger", type = "mwseLogger|nil", description = "The logger, if it exists." }
+	}
+}
```
I'm also not entirely sure how it should be used. 😅  I don't think I've ever found a use-case for it in all my time using this logger API.

Do you think it's worth keeping this method or should I just cut it? I can try to think of some examples if you think it's worth keeping.

C3pa replied:
All right, let's avoid the XY problem. My original thought was: How do we use the new logging API in a multi-file mod?

That's basically the same question as here: https://github.com/MWSE/MWSE/pull/462#discussion_r2047457594.

herbertgithubaccount commented on autocomplete/definitions/namedTypes/mwseLogger/getLevelStr.lua:
```diff
@@ -0,0 +1,7 @@
+	description = [[Gets a `string` representation of the current logging level.]],
+	arguments = {
+		{ name = "level", type = "mwseLogger.LOG_LEVEL",  optional = true, description = "If provided, a string representation of this logging level will be returned. If `nil`, then a string representation of the current logging level will be returned."}
+	}
```
Do you think I should rename it to `getLevelString` or something similar? I'm still not sure how I feel about using `Str` here.

C3pa replied:
I also think `Str` is a bit uncommon in Lua land. Feel free to rename it.

herbertgithubaccount commented on docs/source/guides/logging.md:
````diff
@@ -1,74 +1,240 @@
+```
+becomes
+```lua
+log(formatString, func(...))
````
Very good idea!

herbertgithubaccount commented on docs/source/guides/logging.md:
````diff
@@ -1,74 +1,240 @@
+### Set a custom output file.
+You can ask your logger to record its output in a separate file. To do so, write
+```lua
+log.SetOutputFile("my file")
+```
+This will ensure your log messages get written to `Data Files/MWSE/logs/my file.log`. 
+You can also write
+```lua
+log.SetOutputFile(true)
````
Oops 😅

herbertgithubaccount commented on docs/source/guides/logging.md:
````diff
@@ -1,74 +1,240 @@
+```lua
+log.SetOutputFile(true)
+```
+to automatically generate an output file based on the name of your mod. 
+For example, if your mod was named "My Awesome Mod", then the above code would result in your logging statements
+being written to `Data Files/MWSE/logs/My Awesome Mod.log`.
+
````
Sorry about that. I'm so used to writing in LaTeX that I got used to being able to add a newline after sentences. I didn't realize markdown was different and that it would add a new paragraph after a single line break. I will be more vigilant about that going forward.

C3pa replied:
Yeah, markdown can be unintuitive at times 😃

herbertgithubaccount commented on autocomplete/definitions/namedTypes/mwseLogger/setLogLevel.lua:
I will add it back. Sorry for deleting it.

C3pa reviewed (commented):
A reminder to also rename `LOG_LEVEL` to `logLevel` in the docs.

C3pa commented:
Can you also add this definition as: `autocomplete\definitions\global\mwse\Logger.lua`:
```lua
return {
	type = "value",
	description = [[...]],
	valuetype = "mwseLogger",
}
```
to make this more easily discoverable in the docs.

herbertgithubaccount commented on docs/source/guides/logging.md:
```diff
@@ -1,74 +1,240 @@
 
-## Registering and using your Logger
+1. `ERROR`: Something bad happened.
+2. `WARN`: Something bad _almost_ happened, or something bad _did_ happen but it's not a big deal.
```
Yeah I think so too. I can remove it. I'm not quite sure what to replace it with. What do you think of "... or something bad _did happen, but it's not a major problem"? Basically I want to communicate that this can be used to communicate minor errors.

C3pa replied:
In various articles on the topic of logging levels, they usually describe these log levels as:
* Error: a problem occurred, after which the application can continue operating at a reduced level of functionality
* Warning: a potential issue 

Some examples:
https://stackoverflow.com/a/36656073/26726793
https://stackoverflow.com/a/64806781/26726793 - This diagram sums it up nicely.
https://betterstack.com/community/guides/logging/log-levels-explained/#2-error

In our case, I think you can say "something potentially bad happened" or you can reword it differently as you see fit.

herbertgithubaccount replied:
I updated this section in the latest commit. Let me know what you think about the new version. I'm still on the fence about including "or something bad _almost_ happened", but I think it's helpful to mention that sometimes `WARN` messages are used to communicate that a problem was predicted and mitigated.

C3pa commented on autocomplete/definitions/namedTypes/mwseLogger/setLogLevel.lua:
```diff
@@ -2,6 +2,6 @@ return {
 	type = "method",
```
If you want, now you can add `deprecated = true`.

C3pa commented on autocomplete/definitions/namedTypes/mwseLogger/setLevel.lua:
```diff
@@ -0,0 +1,13 @@
+Use whichever one you prefer.
+]],
+	arguments = {
+		{ name = "newLogLevel", type = "mwseLogger.logLevel" }
```
Since this also supports passing in string representations of log levels, which we don't codify using `---@alias`, could you add a `|string` to the argument type so LuaLS won't show warnings when using the method that way?

C3pa commented on autocomplete/definitions/namedTypes/mwseLogger/setLogLevel.lua:
```diff
@@ -2,6 +2,6 @@ return {
 	description = [[Set the log level. Options are: "TRACE", "DEBUG", "INFO", "WARN", "ERROR" and "NONE".]],
 	arguments = {
-		{ name = "newLogLevel", type = "mwseLoggerLogLevel" }
+		{ name = "newLogLevel", type = "mwseLogger.logLevel" }
```
Can you also add a `|string` for the same reason as above?

herbertgithubaccount commented:
I think I've now implemented all of your suggestions except for the ones pertaining to adding `"|string"` to the documentation files. 

I have instead added an `---@alias` for `mwseLogger.logLevelString` that stores the string representations of the various logging levels. In the places that are fine with either a string or numeric representation, I have changed the type to be `"mwseLogger.logLevel|mwseLogger.logLevelString"`. Let me know if you think this works, but otherwise I can revert the change and instead add `"|string"`.

C3pa reviewed (approved):
I think the introduction of the `mwseLogger.logLevelString` alias is a good call, and even better since it's more strongly-typed than just `|string`.

C3pa commented:
Okay, I'm merging this. I congratulate you on getting your first contribution in 😄!

Now, seriously, this has a lot of changes, but since we added tests that should cover most common ways the API could be used, I don't expect this to break backwards-compatibility with existing mods. If some edge cases come up, these can be patched later.
